framework/src/org/checkerframework/framework/flow/CFAbstractValue.java

```diff
@@ -4,550 +4,481 @@
 import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedTypeVariable;
 import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedWildcardType;
 import org.checkerframework.framework.type.GenericAnnotatedTypeFactory;
 import org.checkerframework.framework.type.QualifierHierarchy;
 import org.checkerframework.framework.type.TypeHierarchy;
-import org.checkerframework.framework.type.visitor.AnnotatedTypeMerger;
 import org.checkerframework.framework.util.AnnotatedTypes;
+import org.checkerframework.framework.util.PluginUtil;
 import org.checkerframework.javacutil.AnnotationUtils;
 import org.checkerframework.javacutil.ErrorReporter;
 import org.checkerframework.javacutil.InternalUtils;
+import org.checkerframework.javacutil.TypesUtils;
 
 /**
- * An implementation of an abstract value used by the Checker Framework org.checkerframework.dataflow
- * analysis. Contains a set of annotations.
+ * An implementation of an abstract value used by the Checker Framework
+ * org.checkerframework.dataflow analysis.
  *
- * @author Stefan Heule
+ * A value holds a set of annotations and a type mirror.  The set of annotations represents the
+ * primary annotation on a type; therefore, the set of annotations must have an annotation for
+ * each hierarchy unless the type mirror is a type variable or a wildcard that extends a type
+ * variable. Both type variables and wildcards may be missing a primary annotation.  For this set
+ * of annotations, there is an additional constraint that only wildcards that extend type
+ * variables can be missing annotations because
+ *
+ * In order to compute {@link #leastUpperBound(CFAbstractValue)} and
+ * {@link #mostSpecific(CFAbstractValue, CFAbstractValue)}, the case where one value has an
+ * annotation in a hierarchy and the other does not must be handled.  For type variables, the
+ * {@link AnnotatedTypeVariable} for the declaration of the type variable is used.  The
+ * {@link AnnotatedTypeVariable} is computed using the type mirror.  For wildcards, it is not
+ * always possible to get the {@link AnnotatedWildcardType} for the type mirror.  However, a
+ * CFAbstractValue's type mirror is only a wildcard if the type of some expression is a wildcard,
+ * so it is always the use of a wildcard rather than the declaration. For uses of wildcards, only
+ * the primary annotation on the upper bound is ever used.  So, the set of annotations represents
+ * the primary annotation on the wildcard's upper bound.  If that upper bound is a type variable,
+ * then the set of annotations could be missing an annotation in a hierarchy.
  *
+ * @author Stefan Heule
  */
 public abstract class CFAbstractValue<V extends CFAbstractValue<V>> implements AbstractValue<V> {
 
     /**
-     * The analysis class this store belongs to.
+     * The analysis class this value belongs to.
      */
     protected final CFAbstractAnalysis<V, ?, ?> analysis;
 
     protected final TypeHierarchy typeHierarchy;
```

Comment:
I think we could safe a lot of memory by not having all of these fields. There will be a lot CFAbstractValue instances and most of these fields will be the same.
Most of them are reachable through `analysis`. Calling the getter methods would safe storage.